### ISSUE TITLE
hight cpu usage v1.2.0

### ISSUE THREAD

**Issue by wuestkamp** [objective-see/LuLu#138]
Hey,

very high cpu usage occasionally, like 90-100%!
OSX High Sierra.

Anyone else experiencing this or tips to solve? I then have to deactivate it.

Cheers,
Kim

**Comment by daslicht** [comment]
any news on this ?

**Comment by daslicht** [comment]
bump

**Comment by heviiguy** [comment]
Yes, @wuestkamp, you're not the only one suffering from this problem. It's been "open" on another thread since February.

**Comment by kdambekalns** [comment]
Seems to be caused by #143 - blocked processes keep running!?

**Comment by agucova** [comment]
I'm on Catalina and Lulu seems to use between 15% and 60% of my CPU at all times.

**Comment by heviiguy** [comment]
My CPU is back to acting normal again. But only after removing all traces of even the most recent version of Lulu!

**Comment by heviiguy** [comment]
## SOLVED!
I couldn't stay away from Lulu. Once again, I reinstalled the program. However, shortly after doing so, my resource usage went haywire again. In response, I disabled Lulu and force-killed it whenever it would auto-magically start. Time for a side-track:

A week or so ago I noticed that I was running into high CPU rates whenever I used dual monitors. After doing a bit of searching, I found others who had experienced similar issues. The common thread was that they, as I was, were running [Übersicht](http://tracesof.net/uebersicht/). With this in mind, I did a bit of fiddling and found that if I only have the Übersicht widgets displayed on the MBP screen, the _computer calmed-down completely_.

This morning I had a "Eureka" moment. I wondered if Übersicht somehow also caused Lulu to suck the living life from my Mac. So, I brought Lulu back from her forced coma and disabled all widgets. Then, one by one, the widgets were enabled until one of them caused the MBP's fan to howl in fury. The offending widget was  "_**disk-usage.widget**_".  Disabling it brought everything back to normal.

I'm finally a Happy Camper!

Bonus: I now have both Lulu and the formerly-corrupt Übersicht widget operating! All it took was a simple refresh rate change in the latter's coffee file.

**Comment by pzrq** [comment]
I hope this is helpful, though I may be totally wrong (skipped OS when doing CompSci at uni a decade ago). Please take with a grain of salt and call out anything I've got wrong.

My curiosity was piqued by LuLu v1.2.3 taking 10-30% CPU on one machine (albeit a slower 2017 13" Core i5) and <1% CPU on another (2017 13" Core i7).

Nevertheless, I'm seeing 2.6s taken when using Activity Monitor > "i" icon > Sample on the skower machine:

<img width="467" alt="Screen Shot 2020-06-13 at 3 12 02 pm" src="https://user-images.githubusercontent.com/1217010/84560507-4d2a5980-ad88-11ea-8919-fba0967cfeeb.png">
<img width="559" alt="Screen Shot 2020-06-13 at 2 59 56 pm" src="https://user-images.githubusercontent.com/1217010/84560379-c759de80-ad86-11ea-9790-473ba7872aab.png">
<img width="986" alt="Screen Shot 2020-06-13 at 2 56 48 pm" src="https://user-images.githubusercontent.com/1217010/84560381-ca54cf00-ad86-11ea-8170-5ed0112e5641.png">

As noted on https://github.com/objective-see/LuLu/issues/183#issuecomment-604286718 Apple is deprecating some of these, so for example [this Darwin Kernel Programming doc on Mach](https://developer.apple.com/library/archive/documentation/Darwin/Conceptual/KernelProgramming/Mach/Mach.html) states:

> There are two fundamentally different Mach APIs for raw manipulation of ports—the `mach_ipc` family and the `mach_msg` family. Within reason, both families may be used with any IPC object; however, the `mach_ipc` calls are preferred in new code. The `mach_ipc` calls maintain state information where appropriate in order to support the notion of a transaction. The `mach_msg` calls are supported for legacy code but deprecated; they are stateless. 

Sounds like finding a way to replace the `mach_msg` calls with `mach_ipc` will help resolve #183 / #200 as a bonus, but that looks to be buried inside things like [`KSCrash`](https://github.com/kstenerud/KSCrash) and [`NSRunLoop`](https://github.com/objective-see/LuLu/search?q=NSRunLoop&unscoped_q=NSRunLoop&type=Code). No idea who's responsible for that code, for example this [Apple NSRunLoop doc](https://developer.apple.com/documentation/foundation/nsrunloop) doesn't suggest it's deprecated so may be up to Apple to fix at least part of it?

**Comment by pzrq** [comment]
Ok, not sure why but after ~20 mins of CPU time, LuLu 1.2.3 has decided to drop back down to what I observed on my other machine, so whatever LuLu is actually doing, in my case it didn't go forever:

<img width="475" alt="Screen Shot 2020-06-13 at 5 10 39 pm" src="https://user-images.githubusercontent.com/1217010/84562616-0bee7580-ad99-11ea-89a1-daa0b6b414ea.png">
